eZ/Publish/Core/Persistence/Legacy/TransactionHandler.php

```diff
@@ -0,0 +1,113 @@
+<?php
+/**
+ * File containing the Legacy Storage TransactionHandler
+ *
+ * @copyright Copyright (C) 1999-2014 eZ Systems AS. All rights reserved.
+ * @license http://www.gnu.org/licenses/gpl-2.0.txt GNU General Public License v2
+ * @version //autogentag//
+ */
+
+namespace eZ\Publish\Core\Persistence\Legacy;
+
+use eZ\Publish\SPI\Persistence\TransactionHandler as TransactionHandlerInterface;
+use eZ\Publish\Core\Persistence\Legacy\Content\Type\MemoryCachingHandler as CachingContentTypeHandler;
+use eZ\Publish\Core\Persistence\Legacy\Content\Language\CachingHandler as CachingLanguageHandler;
+use eZ\Publish\Core\Persistence\Database\DatabaseHandler;
+use Exception;
+use RuntimeException;
+
+/**
+ * The Transaction handler for Legacy Storage Engine
+ *
+ * @since 5.3
+ */
+class TransactionHandler implements TransactionHandlerInterface
+{
+    /**
+     * @var \eZ\Publish\Core\Persistence\Database\DatabaseHandler
+     */
+    protected $dbHandler;
+
+    /**
+     * @var \eZ\Publish\SPI\Persistence\Content\Type\Handler
+     */
+    protected $contentTypeHandler;
+
+    /**
+     * @var \eZ\Publish\SPI\Persistence\Content\Language\Handler
+     */
+    protected $languageHandler;
+
+    /**
+     * @param \eZ\Publish\Core\Persistence\Database\DatabaseHandler $dbHandler
+     * @param \eZ\Publish\Core\Persistence\Legacy\Content\Type\MemoryCachingHandler $contentTypeHandler
+     * @param \eZ\Publish\Core\Persistence\Legacy\Content\Language\CachingHandler $languageHandler
+     */
+    public function __construct(
+        DatabaseHandler $dbHandler,
+        CachingContentTypeHandler $contentTypeHandler = null,
+        CachingLanguageHandler $languageHandler = null
+    )
+    {
+        $this->dbHandler = $dbHandler;
+        $this->contentTypeHandler = $contentTypeHandler;
+        $this->languageHandler = $languageHandler;
+    }
+
+    /**
+     * Begin transaction
+     */
+    public function beginTransaction()
+    {
+        $this->dbHandler->beginTransaction();
+    }
+
+    /**
+     * Commit transaction
+     *
+     * Commit transaction, or throw exceptions if no transactions has been started.
+     *
+     * @throws \RuntimeException If no transaction has been started
+     */
+    public function commit()
+    {
+        try
+        {
+            $this->dbHandler->commit();
+        }
+        catch ( Exception $e )
+        {
+            throw new RuntimeException( $e->getMessage() );
+        }
+    }
+
+    /**
+     * Rollback transaction
+     *
+     * Rollback transaction, or throw exceptions if no transactions has been started.
+     *
+     * @throws \RuntimeException If no transaction has been started
+     */
+    public function rollback()
+    {
+        try
+        {
+            $this->dbHandler->rollback();
+
+            // Clear all caches after rollback
```

Comment:
Well, I'm guessing it's here 'cause cache might have been warmed up for created elements, and might therefore contain invalid caches.

It would a bit brutal, but not unnecessary...